src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java

```diff
@@ -621,6 +621,45 @@ public static int typeDeclarationNameMatchingCount(String patternClass,
         return result;
     }
 
+    /**
+     * Calculates and returns the type declaration matching count when {@code classToBeMatched} is
+     * considered to be super class of an anonymous inner class.
+     *
+     * <p>
+     * Suppose our pattern class is {@code Main.j} and class to be matched is
+     * {@code Main.j.a.b} then type declaration name matching count would be calculated by
+     * comparing every character, and updating main counter when we hit "." or when it is the last
+     * character of the pattern class and certain conditions are met. This is done so that matching
+     * count is 6 instead of 5. This is due to the fact that pattern class can contain anonymous
+     * inner class object of a nested class which isn't true in case of extending classes as you
+     * can't extend nested classes.
+     * </p>
+     *
+     * @param patternTypeDeclaration type declaration against which the given type declaration has
+     *                               to be matched
+     * @param typeDeclarationToBeMatched type declaration to be matched
+     * @return type declaration matching count
+     */
+    public static int getAnonSuperTypeDeclarationMatchingCount(String patternTypeDeclaration,
+                                                               String typeDeclarationToBeMatched) {
+        final int minLength = Math
+            .min(typeDeclarationToBeMatched.length(), patternTypeDeclaration.length());
+        int result = 0;
+        final boolean shouldCountBeUpdatedAtLastCharacter =
+            typeDeclarationToBeMatched.length() > minLength
+                && typeDeclarationToBeMatched.charAt(minLength) == PACKAGE_SEPARATOR;
+
+        for (int i = 0; i < minLength
+            && patternTypeDeclaration.charAt(i) == typeDeclarationToBeMatched.charAt(i); ++i) {
+            if (i == minLength - 1
+                && shouldCountBeUpdatedAtLastCharacter
+                || patternTypeDeclaration.charAt(i) == PACKAGE_SEPARATOR) {
```

Comment:
it is better now, but there is a still duplication that can be avoided
what about
```
        final boolean shouldCountBeUpdatedAtLastCharacter =
            typeDeclarationToBeMatched.length() > minLength;
// in for loop
if (patternTypeDeclaration.charAt(i) == PACKAGE_SEPARATOR && (i != minLength - 1
                || shouldCountBeUpdatedAtLastCharacter) {
```
in this case there is no duplication.

Also I do not like that we compare `typeDeclarationToBeMatched` length with min length, seems a bit odd.

Reply:
> it is better now, but there is a still duplication that can be avoided

Where is it, I am unable to see it (the condition is the same, it is showing outdated as I changed `i` -> `index`).

> Also I do not like that we compare typeDeclarationToBeMatched length with min length, seems a bit odd.

There are cases when its length is less than min length, the condition is necessary for avoiding StringIndexOutOfBoundsException

Reply:
I will respond a bit later, it requires some long description)

Reply:
Yes, we can't have two consecutive package separators but the condition is not redundant, this is the only case where we need not evaluate 
`patternTypeDeclaration.charAt(index) == packageSeparator` as the condition before `||` will be true, but in all other cases, this condition is important, I initially had some pitest failures but after looking closely, it was due to missing examples and not redundant conditions.